staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cache_simple.go

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import "sync"
+
+// simple key/value cache with read/write locking
+type simpleCache struct {
+	lock *sync.RWMutex
+	data map[string]*cacheRecord
+}
+
+func newSimpleCache() cache {
+	return &simpleCache{
+		lock: &sync.RWMutex{},
+		data: map[string]*cacheRecord{},
+	}
+}
+func (c *simpleCache) get(key string) (*cacheRecord, bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	value, ok := c.data[key]
+	return value, ok
+}
+func (c *simpleCache) set(key string, value *cacheRecord) *cacheRecord {
```

Comment:
If `failureTTL` is non-zero, this will caching failed attempts. This would allow DoS by filling all available memory with the underlying `map` here. Maybe it should have a maximum size? Maybe even separate bounds for successful and unsuccessful records?